### PR TITLE
Increase fish number and make its change easier

### PR DESCRIPTION
**Pull request by gyagp**
To cater the powerful desktops, fish number is creased so that FPS won't
hit the roof 60 and an algorithm is introduced to decide the numbers
among small, medium and big fishes. Also to make future change of fish
number easier, the relevant logic is moved from html to js.

### CODE REVIEW

**Comment by gyagp** [comment]
With more and more powerful desktop hardware and browsers, more fish number is required so that 60FPS will not be hit. More detailed test data can be found at crbug.com/755897. This PR is per requests from @kenrussell and some internal needs of Intel for web performance test. Please review!

File: aquarium/aquarium.js [hunk 1]
```diff
@@ -854,8 +841,43 @@ function initialize() {
   Log("--Setup Laser----------------------------------------");
   var laser = setupLaser();
 
+  num = [1, 100, 500, 1000, 5000, 10000, 15000, 20000, 25000, 30000];
```
Comment (greggman) [inline aquarium/aquarium.js]: var?

File: aquarium/aquarium.js [hunk 2]
```diff
@@ -854,8 +841,43 @@ function initialize() {
...
+    div.innerHTML = num[i];
+    parentElem.insertBefore(div, changeViewElem);
+  }
+
   for (var ff = 0; ff < g_fishTable.length; ++ff) {
     g_fishTable[ff].fishData = [];
+    g_fishTable[ff].num = [];
+  }
+
+  type = ["Big", "Medium", "Small"];
```
Comment (greggman) [inline aquarium/aquarium.js]: var?

File: aquarium/aquarium.js [hunk 3]
```diff
@@ -854,8 +841,43 @@ function initialize() {
...
+  for (var i = 0; i < num.length; ++i) {
+    numLeft = num[i];
```
Comment (greggman) [inline aquarium/aquarium.js]: var?

File: aquarium/aquarium.js [hunk 4]
```diff
@@ -854,8 +841,43 @@ function initialize() {
...
+      for (var ff = 0; ff < g_fishTable.length; ++ff) {
+        var fishInfo = g_fishTable[ff];
+        var fishName = fishInfo.name;
+        if (!fishName.startsWith(type[j])) {
+          continue;
+        }
+        if (type[j] == "Big") {
+          numType = Math.min(numLeft, num[i] < 100 ? 1 : 2);
+        } else if (type[j] == "Medium") {
+          numType = Math.min(numLeft, parseInt(num[i] / 10));
```
Comment (greggman) [inline aquarium/aquarium.js]: is `parseInt` needed?  Also at 5000 fish there will be 500 medium fish where as before there would have only been 160. I'm not sure it matters

Comment (gyagp) [reply aquarium/aquarium.js]: parseInt is needed here as fish number may not be multiple of 10, and I hope the number can be easily changed to any value in future. The algorithm of medium fishes is tuned a bit to limit the total number. Is this more reasonable to you?
Other comments have been addressed, and I uploaded a new PR. 
Thanks for the review!

Comment (greggman) [reply aquarium/aquarium.js]: If you want an int use 

     Math.min(numLeft, num[i] / 10 | 0);

Comment (gyagp) [reply aquarium/aquarium.js]: My bad, and parseInt should be used to convert string to integer. Changed to your suggestion. PTAL!

**Comment by kenrussell** [comment]
Thanks Yang for putting this together and Gregg for reviewing and merging it! It will help a lot with performance tests, where the Aquarium is one of the standard benchmarks at this point.

**Comment by greggman** [comment]
Well since we're on that topic. Are there other benchmarks? Should I look into optimizing the aquarium at all? 

One thing I've noticed is native javascript arrays are *much* faster than typed arrays. Since the start of WebGL I (and I think others) assumed using Float32Arrays for matrix and vector math in JavaScript would be a win. But looking at benchmarks it's not even close. Native Arrays are 20% to 40% faster than Float32Arrays

http://greggman.github.io/webgl-matrix-benchmarks/matrix_benchmark.html

* glMatrix-native is using native arrays
* glMatrix is using Float32Arrays
* twgl-native is using native arrays
* twgl is using Float32Arrays

Ignore TDLFast (Float32Array) vs TDLMath (native) as that TDLMath is old and generic and therefore slow.

Also while we're at it current Firefox is 2x faster than Chrome at these benchmarks 😉 😅

With the added conversion from double to float to call gl.uniform I'm not sure how much faster the aquarium would go but the aquarium is doing one matrix multiply, one matrix inverse, and one matrix transpose per fish, all of which are 20% to 40% faster using native arrays where as the aquarium is currently using Float32Arrays

**Comment by gyagp** [comment]
As to WebGL benchmark, one question is often asked: How performant is WebGL when comparing to its counterpart native application? So ideally, we need both WebGL and native benchmarks to implement same functionalities. And for each benchmark, we need to use latest technologies for it. That is, for native, we can use latest OpenGL/D3D11, and Vulkan/D3D12/Metal. And for WebGL, we need to use WebGL 1, WebGL 2 and WebGPU later on. We may write WebGL benchmark manually, or have it converted from some tools like Emscripten. 
The WebGL will have more and more adoption when its performance gets closer with native application.

**Comment by kenrussell** [comment]
@greggman thanks for pointing out the performance gap between native JavaScript arrays and typed arrays. I've emailed the V8 team about it and will post a pointer to a bug tracking the issue once it's either filed or re-raised. (There are a lot of open performance bugs against typed arrays right now and it's hard to see whether there's already one covering this.)

@gyagp more benchmarks that allow comparison between native applications and the same thing running on the web are definitely needed. WebAssembly is a good technology for comparing native libraries and engines like https://github.com/floooh/oryol with web counterparts. It will be good if we can recode some of these in efficient JavaScript, too.

**Comment by greggman** [comment]
I should point out one more thing about perf in V8 is that context is important. A global function is significantly faster than a closed function even if the code references nothing in the closure. I think I filed bugs on that but didn't see any attention so haven't yet posted more concrete tests. Maybe I can find time to write some.

**Comment by kenrussell** [comment]
@greggman filed https://bugs.chromium.org/p/v8/issues/detail?id=7030 about the performance gap between native and typed arrays, and between Chrome and Firefox.

Regarding additional benchmarks, your Aquarium seems useful in its current form, as several browsers are working on optimizing it. On the other hand if you can think of other pages/apps you or colleagues have built which you think should be running faster, please, highlight them.

I scanned the V8 bug database for bugs you'd filed and didn't see anything regarding performance of functions depending on their nesting / closure context. If you have any findings please do file bugs and let me know about them; I'll help as much as I can to get them looked at. (The V8 team is focusing heavily on performance of real-world sites and JavaScript libraries, and anymore synthetic, CPU-bound benchmarks.)

**Pull request merged**